File: sdk/python/examples/09b_hitl_with_feedback.py
# Copyright (c) 2025 Agentspan
# Licensed under the MIT License. See LICENSE file in the project root for details.

"""Human-in-the-Loop with Custom Feedback.

Demonstrates the general-purpose `respond()` API.  Instead of a binary
approve/reject, the human can send arbitrary feedback that the LLM
processes on its next iteration.

Use case: a content-publishing agent writes a blog post, and a human
editor can approve, reject, or provide revision notes.  The agent
incorporates the feedback and tries again.

Requirements:
    - Conductor server with LLM support
    - AGENTSPAN_SERVER_URL=http://localhost:8080/api as environment variable
    - AGENTSPAN_LLM_MODEL=openai/gpt-4o-mini as environment variable
"""

from agentspan.agents import Agent, AgentRuntime, EventType, tool
from settings import settings


@tool(approval_required=True)
def publish_article(title: str, body: str) -> dict:
    """Publish an article to the blog. Requires editorial approval."""
    return {"status": "published", "title": title, "url": f"/blog/{title.lower().replace(' ', '-')}"}


agent = Agent(
    name="writer",
    model=settings.llm_model,
    tools=[publish_article],
    instructions=(
        "You are a blog writer. When asked to write about a topic, draft an article "
        "and publish it using the publish_article tool. If you receive editorial "
        "feedback, revise the article and try publishing again."
    ),
)


if __name__ == "__main__":
    with AgentRuntime() as runtime:
        # Deploy to server. CLI alternative (recommended for CI/CD):
        #   agentspan deploy examples.09b_hitl_with_feedback
        runtime.deploy(agent)
        runtime.serve(agent)

        # Quick test: uncomment below (and comment out serve) to run directly.
        # result = runtime.stream(
        #     agent, "Write a short blog post about the benefits of code review"
        # )
        # print(f"Workflow started: {result.workflow_id}\n")

        # for event in result:
        #     print(f'event type: {event.type} --> {event.content}')
        #     if event.type == EventType.THINKING:
        #         print(f"  [thinking] {event.content}")

        #     elif event.type == EventType.TOOL_CALL:
        #         print(f"  [tool_call] {event.tool_name}")
        #         if event.args:
        #             title = event.args.get("title", "")
        #             body = event.args.get("body", "")
        #             if title:
        #                 print(f"    Title: {title}")
        #             if body:
        #                 preview = body[:200] + "..." if len(body) > 200 else body
        #                 print(f"    Body:  {preview}")

        #     elif event.type == EventType.GUARDRAIL_FAIL:
        #         print(f"  [guardrail failed] {event.guardrail_name}")
        #         if event.args:
        #             title = event.args.get("title", "")
        #             body = event.args.get("body", "")
        #             if title:
        #                 print(f"    Title: {title}")
        #             if body:
        #                 preview = body[:200] + "..." if len(body) > 200 else body
        #                 print(f"    Body:  {preview}")

        #     elif event.type == EventType.TOOL_RESULT:
        #         print(f"  [tool_result] {event.tool_name} -> {event.result}")

        #     elif event.type == EventType.WAITING:
        #         print(f"\n--- Editorial Review Required ---")
        #         print("  [a] Approve and publish")
        #         print("  [r] Reject entirely")
        #         print("  [f] Provide feedback for revision")
        #         print()

        #         choice = input("  Choice (a/r/f): ").strip().lower()

        #         if choice == "a":
        #             result.approve()
        #             print("  Approved for publication!\n")
        #         elif choice == "r":
        #             reason = input("  Rejection reason: ").strip()
        #             result.reject(reason or "Does not meet editorial standards")
        #             print("  Rejected.\n")
        #         else:
        #             feedback = input("  Feedback: ").strip()
        #             result.respond({"feedback": feedback})
        #             print("  Feedback sent, agent will revise...\n")

        #     elif event.type == EventType.ERROR:
        #         print(f"  [error] {event.content}")

        #     elif event.type == EventType.DONE:
        #         print(f"\n  [done] {event.output}")

        # # Access the structured result after streaming
        # final = result.get_result()
        # print(f"\nTool calls made: {len(final.tool_calls)}")
        # print(f"Status: {final.status}")

File: sdk/typescript/examples/09b-hitl-with-feedback.ts
/**
 * Human-in-the-Loop with Custom Feedback.
 *
 * Demonstrates the general-purpose respond() API. Instead of a binary
 * approve/reject, the human can send arbitrary feedback that the LLM
 * processes on its next iteration.
 *
 * Use case: a content-publishing agent writes a blog post, and a human
 * editor can approve, reject, or provide revision notes. The agent
 * incorporates the feedback and tries again.
 *
 * Requirements:
 *   - Conductor server with LLM support
 *   - AGENTSPAN_SERVER_URL=http://localhost:6767/api as environment variable
 *   - AGENTSPAN_LLM_MODEL=openai/gpt-4o-mini as environment variable
 */

import { z } from 'zod';
import { Agent, AgentRuntime, tool } from '../src/index.js';
import { llmModel } from './settings.js';

const publishArticle = tool(
  async (args: { title: string; body: string }) => {
    return {
      status: 'published',
      title: args.title,
      url: `/blog/${args.title.toLowerCase().replace(/ /g, '-')}`,
    };
  },
  {
    name: 'publish_article',
    description: 'Publish an article to the blog. Requires editorial approval.',
    inputSchema: z.object({
      title: z.string().describe('Article title'),
      body: z.string().describe('Article body'),
    }),
    approvalRequired: true,
  },
);

export const agent = new Agent({
  name: 'writer',
  model: llmModel,
  tools: [publishArticle],
  instructions:
    'You are a blog writer. When asked to write about a topic, draft an article ' +
    'and publish it using the publish_article tool. If you receive editorial ' +
    'feedback, revise the article and try publishing again.',
});

// Only run when executed directly (not when imported for discovery)
async function main() {
  const runtime = new AgentRuntime();
  try {
    // Deploy to server. CLI alternative (recommended for CI/CD):
    //   agentspan deploy <module>
    // await runtime.deploy(agent);
    // await runtime.serve(agent);
    const result = await runtime.run(agent, 'Write a short blog post outline about the benefits of code review. Do not publish it.');
    result.printResult();

    // Production pattern:
    // 1. Deploy once during CI/CD:
    // await runtime.deploy(agent);
    // CLI alternative:
    // agentspan deploy --package sdk/typescript/examples
    //
    // 2. In a separate long-lived worker process:
    // await runtime.serve(agent);

    // Interactive HITL alternative:
    // const runtime = new AgentRuntime();
    // try {
    // const streamHandle = await runtime.stream(
    // agent,
    // 'Write a short blog post about the benefits of code review',
    // );
    // console.log(`Execution started: ${streamHandle.executionId}\n`);

    // for await (const event of streamHandle) {
    // console.log(`event type: ${event.type} --> ${event.content}`);

    // switch (event.type) {
    // case 'thinking':
    // console.log(`  [thinking] ${event.content}`);
    // break;

    // case 'tool_call': {
    // console.log(`  [tool_call] ${event.toolName}`);
    // if (event.args) {
    // const args = event.args as Record<string, string>;
    // if (args.title) {
    // console.log(`    Title: ${args.title}`);
    // }
    // if (args.body) {
    // const preview =
    // args.body.length > 200
    // ? args.body.slice(0, 200) + '...'
    // : args.body;
    // console.log(`    Body:  ${preview}`);
    // }
    // }
    // break;
    // }

    // case 'guardrail_fail': {
    // console.log(`  [guardrail failed] ${event.guardrailName}`);
    // if (event.args) {
    // const args = event.args as Record<string, string>;
    // if (args.title) {
    // console.log(`    Title: ${args.title}`);
    // }
    // if (args.body) {
    // const preview =
    // args.body.length > 200
    // ? args.body.slice(0, 200) + '...'
    // : args.body;
    // console.log(`    Body:  ${preview}`);
    // }
    // }
    // break;
    // }

    // case 'tool_result':
    // console.log(
    // `  [tool_result] ${event.toolName} -> ${JSON.stringify(event.result)}`,
    // );
    // break;

    // case 'waiting':
    // console.log(`\n--- Editorial Review Required ---`);
    // console.log('  [a] Approve and publish');
    // console.log('  [r] Reject entirely');
    // console.log('  [f] Provide feedback for revision');
    // console.log();
    // // Auto-approve since we can't do interactive stdin
    // console.log('  Auto-approving for demo...');
    // await streamHandle.approve();
    // console.log('  Approved for publication!\n');
    // break;

    // case 'error':
    // console.log(`  [error] ${event.content}`);
    // break;

    // case 'done':
    // console.log(`\n  [done] ${JSON.stringify(event.output)}`);
    // break;
    // }
    // }

    // // Access the structured result after streaming
    // const final = await streamHandle.getResult();
    // console.log(`\nTool calls made: ${final.toolCalls.length}`);
    // console.log(`Status: ${final.status}`);
  } finally {
    await runtime.shutdown();
    // }
}

if (process.argv[1]?.endsWith('09b-hitl-with-feedback.ts') || process.argv[1]?.endsWith('09b-hitl-with-feedback.js')) {
  main().catch(console.error);
}
